tests/KubernetesClient.Tests/Mock/MockKubeApiServer.cs

```diff
@@ -10,26 +11,206 @@
+        private readonly MockKubeServerFlag[] _serverFlags;
+        private readonly string _mockAddedEventStreamLine = BuildWatchEventStreamLine(WatchEventType.Added);
+        private readonly string _mockDeletedStreamLine = BuildWatchEventStreamLine(WatchEventType.Deleted);
+        private readonly string _mockModifiedStreamLine = BuildWatchEventStreamLine(WatchEventType.Modified);
+        private readonly string _mockErrorStreamLine = BuildWatchEventStreamLine(WatchEventType.Error);
+        private bool _disposed;
+        private const string MockBadStreamLine = "bad json";
 
-        public MockKubeApiServer(ITestOutputHelper testOutput, Func<HttpContext, Task<bool>> shouldNext = null,
-            Action<ListenOptions> listenConfigure = null,
+        public MockKubeApiServer(ITestOutputHelper testOutput, params MockKubeServerFlag[] serverFlags)
+        {
+            _serverFlags = serverFlags;
+            _webHost = BuildWebHost(testOutput, ShouldNext, null, null); // to avoid making ShouldNext static
+            _webHost.Start();
+        }
+
+        public MockKubeApiServer(
+            ITestOutputHelper testOutput,
+            Func<HttpContext, Task<bool>> shouldNext = default,
+            Action<ListenOptions> listenConfigure = default,
             string resp = MockPodResponse)
+        {
+            _webHost = BuildWebHost(testOutput, shouldNext, listenConfigure, resp);
+            _webHost.Start();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            // Execute if resources have not already been disposed.
+            if (!_disposed)
+            {
+                // If the call is from Dispose, free managed resources.
+                if (disposing)
+                {
+                    _webHost.StopAsync();
+                    _webHost.WaitForShutdown();
+                    _webHost.Dispose();
+                }
+            }
+
+            _disposed = true;
+        }
+
+        private static string BuildWatchEventStreamLine(WatchEventType eventType)
+        {
+            var corev1PodList = JsonConvert.DeserializeObject<V1PodList>(MockPodResponse);
+            return JsonConvert.SerializeObject(
+                new Watcher<V1Pod>.WatchEvent { Type = eventType, Object = corev1PodList.Items.First() },
+                new StringEnumConverter());
+        }
+
+        private async Task WriteStreamLine(HttpContext httpContext, string reponseLine)
+        {
+            const string crlf = "\r\n";
+            await httpContext.Response.WriteAsync(reponseLine.Replace(crlf, "")).ConfigureAwait(false);
+            await httpContext.Response.WriteAsync(crlf).ConfigureAwait(false);
+            await httpContext.Response.Body.FlushAsync().ConfigureAwait(false);
+        }
+
+        private async Task<bool> ShouldNext(HttpContext httpContext)
```

Comment:
not quite get the idea of `shouldnext` here
`shouldnext` is a func (closure) can be defined anywhere. 

here i think you are doing your test specific `shouldnext`, why not pass it as a param?

Reply:
I was just trying to clean up my test code and make it a little more readable by using flags instead. If you feel this it adds too too much in the Mock server I can move it all out.

Reply:
IMHO, keep all test behavior as a param to `shouldnext` would be more readable

for example
```
// test throw500
shouldnext(() => throw 500)
```
otherwise, have to
1. add a flag
1. add case
1. call with flag param

why not put them together